### ISSUE TITLE
testing-library/no-node-access - recommended way to deal with children nodes / classnames

### ISSUE THREAD

**Issue by rdelgatte** [testing-library/eslint-plugin-testing-library#380]
What is the recommended way to deal with children nodes / to look for a specific item by its classname and testing library rules? 

Here is an example of something I wrote earlier and I can't figure out how to fit the rules expectations to get the same result. 

```ts
const paragraphs = screen.getByRole('article').children;
```

Also, what about looking for html elements by their classname like using `getElementsByClassName`, is there any workaround to do that by respecting the `testing-library/no-node-access` rule?

Thank you for your help!

**Comment by Belco90** [comment]
Hey @rdelgatte! This is a tricky question.

First of all, it's really difficult to suggest "how to query elements for passing your tests" without seeing the actual test. In order to avoid using `children` or `getElementsByClassName` you could need to change the way your test is implemented.

Secondly, the purpose of the rule is actually to force you to change the way you query the elements so it's not tied to implementation but to behavior. For example, when you are querying the paragraphs like `screen.getByRole('article').children` you are tied to its implementation since the paragraphs must be direct children from `article` tag. Depending on what you want to assert, you could [query them by their text](https://testing-library.com/docs/queries/bytext) only for those [within an `article`](https://testing-library.com/docs/dom-testing-library/api-within/) so you make sure they are under the expected structure, but not necessarily direct children.

If you really need to check they are direct children of an article, or query elements by class name because then I think the best option would be just to disable the rule. I'd try to find a different approach to query the elements you want if not.

Does that make sense to you?

**Comment by rdelgatte** [comment]
Hello @Belco90

Thank you for your quick and relevant feedback, this definitely helps understanding the mindset here (not relying over the implementation) and I also try to challenge the way the assertions are done. I've been relying over a property-based testing library to generate random datasets (like a list of articles with randomly-generated attributes values), that's why I can't rely over some "static" text values.

I was just wondering if you had another way of doing it in a more "testing-library-oriented" mindset. 

Anyway, I feel like there shouldn't be an easy way to deal with that (which makes sense) so I am closing this issue 👍

**Comment by Belco90** [comment]
Yeah, I'm afraid there is no particular mindset around that apart from changing the approach of your asserts. Anyway, a couple of extra notes:

- [`jest-dom`](https://github.com/testing-library/jest-dom) could make it easier for you to assert the desired elements. Matchers like `toContainElement` or `toContainHTML` might be helpful for you.
- Even if generate random datasets, you should be able to know them to query the corresponding element by text right? That would be the first thing I'd try to change in the approach if you want a more Testing Library-ish mindset. Ideally, no random data generated, but I don't have the proper context for it in your codebase of course.

**Comment by alexey-kozlenkov** [comment]
Hey @Belco90 👋 
Perhaps you could advise me on smth similar: how to correctly abide `no-node-access` rule with snapshot tests?

My tests test if some gradient component has correct colors with the help of snapshots.
```describe('AreaGradientComponent', () => {
  const green = '#0f0'
  const red = '#f00'

  const renderAreaGradient = (min: number, max: number) =>
    render(
      <AreaGradientComponent
        id="ID"
        minValue={min}
        maxValue={max}
        positiveColor={green}
        negativeColor={red}
      />,
    )

  it('returns a positive colored component', () => {
    const { container } = renderAreaGradient(0, 100)

    expect(container.firstChild).toMatchSnapshot()
  })  
})
```

`container.firstChild` above is reported with `no-node-acess` rule, however it kinda does not violates [docs](https://testing-library.com/docs/react-testing-library/api#:~:text=Tip%3A%20To%20get%20the%20root%20element%20of%20your%20rendered%20element%2C%20use%20container.firstChild.):

Any suggestions how can I change things around here?

**Comment by Belco90** [comment]
Hey @alexey-kozlenkov! I IMHO you could match the snapshot against the `container` itself, but I think `container.firstChild` is a valid use case for snapshots indeed.

Could you move this to a new ticket, so we can discuss it further and use it for adding a new option to the rule for this case if needed?

Thanks!
